### PR TITLE
Add a new mutator (re-)computing background metric variables on DG grid

### PR DESCRIPTION
**Pull request by yoonso0-0**
This replaces (and renames) the `evolution::initialization::GrTagsForHydro` action, and also can be used for both evolution and initialization phases. Many lines of the code are similar with the original `GrTagsForHydro` action.

Another PR for subcell will be submitted after testing a TOV star on a moving mesh.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/Actions/BackgroundGrVars.hpp [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+ *  - Removes : nothing
+ *  - Modifies : nothing
+ *
+ */
+template <typename System, bool used_in_evolution_phase>
+struct BackgroundGrVars {
+  using simple_tags_from_options = tmpl::list<::Tags::Time>;
+
+  // Collect all the GR quantities used in the templated evolution system
+  using gr_variables_tag = ::Tags::Variables<tmpl::remove_duplicates<
```
Comment (yoonso0-0) [inline src/Evolution/Actions/BackgroundGrVars.hpp]: This line (collection of _all_ the GR related tags) may be unnecessary, would be happy to get comments on this.

File: src/Evolution/Actions/BackgroundGrVars.hpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+ private:
+  template <typename DbTagsList, typename T>
+  static void impl(const gsl::not_null<db::DataBox<DbTagsList>*> box,
+                   const T& solution_or_data) {
+    static constexpr size_t Dim = System::volume_dim;
+
+    using GrVars = typename gr_variables_tag::type;
+
+    if constexpr (used_in_evolution_phase) {
+      // Check if the mesh is actually moving i.e. block coordinate map is
```
Comment (kidder) [inline src/Evolution/Actions/BackgroundGrVars.hpp]: what about the case where the solution is time-dependent, but the mesh is not moving?

Comment (yoonso0-0) [reply src/Evolution/Actions/BackgroundGrVars.hpp]: That's a good point. Can we check if a solution is time-dependent?

Comment (kidder) [reply src/Evolution/Actions/BackgroundGrVars.hpp]: hmm, I don't see a way quickly browsing.  For now just add a warning to the documentation of the action that it assumes the solution is time-independent, and create an issue to remind us to fix this at some point.  Probably doesn't make sense to do that until we can create initial data from the input file for all systems.

Comment (yoonso0-0) [reply src/Evolution/Actions/BackgroundGrVars.hpp]: Okay. I also think this would not be a serious practical problem unless we want to use time-dependent GR solutions as a background, which I can't think of many for now?

File: src/Evolution/Actions/BackgroundGrVars.hpp [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+ *  - Adds:
+ *    * system::spacetime_variables_tag
+ *    * system::flux_spacetime_variables_tag
+ *    * system::inverse_spatial_metric_tag
+ *  - Removes : nothing
+ *  - Modifies : nothing
+ *
+ */
+template <typename System, bool used_in_evolution_phase>
+struct BackgroundGrVars {
```
Comment (nilsdeppe) [inline src/Evolution/Actions/BackgroundGrVars.hpp]: A few questions for our meeting later:

1. I think you can get rid of the `used_in_evolution_phase` and just check if the size of the variables are non-zero.
2. This seems to only be updating the active tags, but always using the GR coordinates. Naively I expect this to not work with subcell. I _think_ what works is if you forward declare the subcell tags and then do the `if constexpr (tag_is_retrievable) {` that'll work with just a forward decl. If not, we (I) need to figure out how to make that work, because it should. Then you can have an internal check basically for "Could I be doing subcell? If so, then check what the active grid is, get the right coordinates, etc. else just do the DG case"

So as-is, I expect the code to actually not work correctly because 1) the face tags aren't updated and 2) the cell-centered tags are always set to the DG grid. I think this code here is different from what you were sharing in our call, though.

**Comment by yoonso0-0** [comment]
@nilsdeppe @kidder I changed the implementation of this from an action to a mutator.

File: src/Evolution/DiscontinuousGalerkin/BackgroundGrVars.hpp [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+/*!
+ * \brief Allocate or assign background general relativity quantities needed
+ * for evolution systems run on a curved spacetime without solving Einstein
+ * equations (e.g. ValenciaDivclean, ForceFree).
+ *
+ * \warning This mutator assumes that the GR analytic data or solution
+ * specifying background spacetime metric is time-independent.
+ *
+ */
+template <typename System, typename Metavariables, bool using_runtime_id>
```
Comment (nilsdeppe) [inline src/Evolution/DiscontinuousGalerkin/BackgroundGrVars.hpp]: camel case `UsingRuntimeId`?

File: src/Evolution/DiscontinuousGalerkin/BackgroundGrVars.hpp [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+      const gsl::not_null<GrVars*> background_gr_vars, const double time,
+      const Domain<Dim>& domain,
+      const tnsr::I<DataVector, Dim, Frame::Inertial>& inertial_coords,
+      const Mesh<Dim>& mesh, const Element<Dim>& element,
+      const T& solution_or_data) {
+    const size_t num_grid_pts = mesh.number_of_grid_points();
+
+    if (contains_allocations(*background_gr_vars)) {  // Evolution phase
+#ifdef SPECTRE_DEBUG
+      const size_t gr_vars_size = (*background_gr_vars).number_of_grid_points();
```
Comment (nilsdeppe) [inline src/Evolution/DiscontinuousGalerkin/BackgroundGrVars.hpp]: why not just inline this call since you only use the value once? Then you can get rid of the `ifdef`

File: src/Evolution/Executables/RelativisticEuler/Valencia/EvolveValencia.hpp [hunk 6]
```diff
@@ -122,7 +123,7 @@ struct EvolutionMetavars {
       is_analytic_data_v<initial_data> xor is_analytic_solution_v<initial_data>,
       "initial_data must be either an analytic_data or an analytic_solution");
 
- using eos_base = typename EquationsOfState::get_eos_base<
+  using eos_base = typename EquationsOfState::get_eos_base<
```
Comment (nilsdeppe) [inline src/Evolution/Executables/RelativisticEuler/Valencia/EvolveValencia.hpp:127]: remove indent?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
You can squash immediately

File: src/Evolution/DiscontinuousGalerkin/BackgroundGrVars.hpp [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+  template <typename T>
+  static void apply(
+      const gsl::not_null<GrVars*> background_gr_vars, const double time,
+      const Domain<Dim>& domain,
+      const tnsr::I<DataVector, Dim, Frame::Inertial>& inertial_coords,
+      const Mesh<Dim>& mesh, const Element<Dim>& element,
+      const T& solution_or_data) {
+    const size_t num_grid_pts = mesh.number_of_grid_points();
+
+    if (contains_allocations(*background_gr_vars)) {  // Evolution phase
```
Comment (wthrowe) [inline src/Evolution/DiscontinuousGalerkin/BackgroundGrVars.hpp]: Just a drive-by comment because I saw it in an email, but this seems like an extremely strange thing to branch on here.  I'm guessing you wanted `background_gr_vars->number_of_grid_points() != 0`.

File: src/Evolution/Executables/RelativisticEuler/Valencia/EvolveValencia.hpp (hunk 6 above)
Comment (yoonso0-0) [reply src/Evolution/Executables/RelativisticEuler/Valencia/EvolveValencia.hpp:127]: I believe this is an action of `clang-format` and actually corrects a missing whitespace

**Comment by yoonso0-0** [comment]
Squashed and rebased on develop :)

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
@kidder @wthrowe do either of you want to look to make sure you're happy too?

**Comment by wthrowe** [comment]
I'm happy with how my comment was dealt with.  I don't think I need to do a full review.

**Review by kidder — APPROVED** [review]
